Verify the bootstrap client cert before using it

smarterclayton opened the pull request:
Before the bootstrap client is used, check a number of conditions that
ensure it can be safely loaded by the server. If any of those conditions
are invalid, re-bootstrap the node. This is primarily to force
bootstrapping without human intervention when a certificate is expired,
but also handles partial file corruption.

smarterclayton committed:
Verify the bootstrap client cert before using it

Before the bootstrap client is used, check a number of conditions that
ensure it can be safely loaded by the server. If any of those conditions
are invalid, re-bootstrap the node. This is primarily to force
bootstrapping without human intervention when a certificate is expired,
but also handles partial file corruption.

smarterclayton committed:
An expired certificate is not compatible

If the certificate in the CSR is expired, it's no good to the code.
Error out with the correct message.

smarterclayton commented:
Second commit detects expired CSR status.certificate.  I'm not sure whether we want to give permission for CSRs to delete their own CSRs, but the current logic for server certificate rotation will wedge if the CSR is expired and the node is rebooted.

We should probably either iterate a sequence (add `-%d` to the end and increment) or try to delete.  We probably need the signing controller to purge certificates that have expired for short rotation to be practical.

Was testing with 10 minute expirations.

liggitt commented:
> current logic for server certificate rotation will wedge if the CSR is expired and the node is rebooted

Really? New private key and name should be used for each new CSR… node should be able to just abandon an expired one

smarterclayton commented:
Nothing is regenerating the private key for client or server certs that I see.

smarterclayton commented:
Ok, I saw we changed the server cert to use Start() always.  The code here is still correct - an expired CSR is not compatible.

smarterclayton commented:
/retest

smarterclayton commented:
I found the issue.  Before we start the cert manager we have to perform the initial bootstrap (using the bootstrap kubeconfig).  If we find a valid key on disk, we never regenerate it.  Because the key never changes, after the CSR expires the kubelet will be unable to start (the old key will be used with the same client info, which means we get the same CSR name).

`bootstrap.LoadClientCert` should delete the private key if it fails to get a new CSR for any reason, probably.

liggitt commented:
> bootstrap.LoadClientCert should delete the private key if it fails to get a new CSR for any reason, probably.

@mikedanese wasn't this to be re-entrant if the CSR wasn't satisfied because the manual approval went long or the controller was backed up, and the kubelet timed out and restarted?

smarterclayton commented:
In this case the CSR is expired, so we are out of luck without deleting the
CSR or picking a new name.

On Thu, Sep 28, 2017 at 10:07 PM, Jordan Liggitt <notifications@github.com>
wrote:

> bootstrap.LoadClientCert should delete the private key if it fails to get
> a new CSR for any reason, probably.
>
> @mikedanese <https://github.com/mikedanese> wasn't this to be re-entrant
> if the CSR wasn't satisfied because the manual approval went long or the
> controller was backed up, and the kubelet timed out and restarted?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/53037#issuecomment-333011687>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pxBnwoHZL5JFxEJM_5R0P7XAvvabks5snFDigaJpZM4PjnvB>
> .
>

smarterclayton committed:
Delete the private key for the bootstrap client cert on failure

Ensures that in a crash loop state we can make forward progress by
generating a new key and hence new CSR. If we do not delete the key, an
expired CSR may block startup.

Also more aggressively delete a bad cert path

smarterclayton commented:
Added a third commit - it may be too aggressive for cleanup (always), so I can move it later if necessary (if request CSR exits out)

smarterclayton commented:
Found a new problem - if a client cert is used for creating CSRs and expires, or if a token is used for creating CSRs and it expires, or if the auth mechanism is revoked, the kubelet will just spin forever trying to bootstrap and anything using a client will effectively be dead.  This is exacerbated when the admin sets an aggressive CSR window - the odds of a kubelet pausing / missing a loop / having connection problems can then put the kubelet into a dead state where it never proceeds.

I *think* in general there needs to be some time based guarantee that ensures that the kubelet starts to crash loop when successive rotations fail.  As long as a crash looping kubelet gets at least one chance to request a new cert using whatever credentials are on disk that is sufficient.  It should not be possible for a kubelet to go zombie.

Scenarios for getting a "cert retrieval client" when rotation is on:

1. kubelet has a bootstrap kube config
2. kubelet has no bootstrap kube config, but flags that tell it to use a kubeconfig on disk
3. kubelet has a previous client cert on disk that the cert store tries to use

Failure modes for the "cert retrieval client":

1. bootstrap kube config is expired or has permission revoked
2. flags for kube config point to expired or revoked tokens
3. admin revokes permission on bootstrap kubeconfig
4. the automated mechanism that puts the config on the node dies

Time when the "cert retrieval client" expires:

1. Before the cert manager starts
2. After the cert manager loads it but before the first request
3. After the first request

Types of errors:

1. 403 - cert expired
2. 403 - master rotated its certs
3. 403 - token expired
4. 403 - RBAC removed permission
5. 403 - code upgrade
6. 404 - cert rotation endpoint is off
7. 500 - master is crashing
8. timeout - master is unreachable

Suggestions:

1. provide an upper bound in time for a kubelet cert manager to fail continuously before it exits
2. check the "cert retrieval client" (whether from bootstrap, store, or kube config) and if it is expired, exit quickly

We assume that in a heavy rotation environment clock skew is unacceptable (skew must be << rotation interval).

smarterclayton committed:
If CSR is deleted, exit immediately

No point in waiting

liggitt commented:
Cases 1,2,3 would be 401, not 403

smarterclayton committed:
Collapse duplicate code into pkg/util/csr

There is no reason to duplicate this code into two places.

smarterclayton committed:
Kubelet should exit if the current client cert has expired

The client cert manager uses the most recent cert to request new
certificates. If that certificate is expired, it will be unable to
complete new CSR requests. This commit alters the manager to force
process exit if no further client cert rotation is possible, which
is expected to trigger a restart of the kubelet and either a
re-bootstrap from the bootstrap kubeconfig or a re-read of the
current disk state (assuming that some other agent is managing the
bootstrap configuration).

This prevents the Kubelet from wedging in a state where it cannot make
API calls.

smarterclayton commented:
/retest

smarterclayton commented:
The last commit (which exits after it can no longer fetch a valid cert) may be too aggressive.  It bounds how long a master outage can go on before the cluster goes from coasting (no API updates, but continue to maintain existing pods) to decaying (because the kubelet has no state).  Eventually checkpointing may reduce that impact, but the maximum downtime of an api server bounds how long the cluster goes before it degrades.

smarterclayton commented:
Proposal - if the server is still responsive, restart.  If the server isn't returning a 401, stay up because availability may be compromised.  Basically, stay a zombie as long as it's better than the alternative.

mikedanese commented on pkg/kubelet/certificate/bootstrap/bootstrap.go:
```diff
@@ -82,20 +94,20 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
 	if err != nil {
 		return fmt.Errorf("unable to build bootstrap client cert path: %v", err)
 	}
+	defer func() {
```
why is this moved above RequestNodeCertificate as well? Does that write anything?

smarterclayton replied:
To avoid leaving a cert on disk if we got a key but failed during req node certificate.  The cert is useless without the key (which we delete above).  This makes the flow a lot more predictable - if we fail to attain a cert, delete key and cert and try again.

mikedanese reviewed (commented):
This all looks reasonable to me.

smarterclayton commented:
I'm going to add one more commit that makes the node go into zombie mode if the server is unreachable when we try to request a cert, but die quickly if the cert is up (to give us another chance to bootstrap).  That's assuming the bootstrap lifetime is longer than the cert lifetime, but if the cert lifetime is much greater than bootstrap lifetime this is less of an issue.

smarterclayton committed:
cache.ListWatchUntil should return err.ErrWaitTimeout

Clients shouldn't have to know about watch.ErrWatchClosed, which is
typically a server side decision to close and always means "Timeout" in
this conetxt.

smarterclayton committed:
Have the certificate manager decide if the server is healthy

Prevent a Kubelet from shutting down when the server isn't responding to
us but we cannot get a new certificate. This allows a cluster to coast
if the master is unresponsive or a node is partitioned and their client
cert expires.

smarterclayton commented:
I stuck one commit on at the end - instead of doing request client cert and waiting for an hour, the cert manager now does shorter waits and tracks whether the server is responding at all.  If we get a definitive response from the server, then the cert manager should not exit when the client cert expires and instead keep on in zombie mode.

1. kubelet starts, gets bootstrap client cert
2. gives bootstrap client cert to cert manager, which requests a new node cert
3. cert expires because the server didn't sign in time
4. cert manager knows that it can't sign a new cert, and it sees 401 or valid errors from the server, so it assumes that it can restart and either bootstrap, or signal a crash loop

The alternate for 4 is that the server isn't able to even contact the server because it gets connection refused, timeout, or a generic proxy error, and so the transport code allows the kubelet to keep limping along after its connection has expired in order to not leave its pods completely unmanaged while the masters are down.

mikedanese commented:
I haven't reviewed the last commit yet. It's on my todo list.

smarterclayton committed:
Move pkg/kubelet/util/csr into client-go

Everything else it depends on was already there, and now we have a
somewhat consistent code chain.

smarterclayton commented:
Updated to react to the change to move certificate manager into client-go - kubelet/util/csr is generic and moved with it.  The node part is effectively generic and can be an example to users.

smarterclayton commented:
/retest

mikedanese commented:
/lgtm
/approve no-issue
